### ISSUE TITLE
binaries in toolchain file are wrongly handled

### ISSUE THREAD

**Issue by stsp** [mesonbuild/meson#13223]
Meson insists on adding some binaries
to the cross-toolchain config file. For example
it warns if `strip` is not there.
First problem is that there can be more
than 1 strip name. Say, for x86_32 it can be:
`i686-linux-gnu-strip`, `i386-elf-strip`,
`x86_64-linux-gnu-strip` (this one is bi-arch).
But the attempt to add a list of names,
results in meson making a single command
of all these names! Instead of to search
for the existing one from the list.

Now I can ignore the toolchain config
and pass the same list to `find_program()`
directly, adding some dummy name to the
toolchain config to avoid the warning.
I also need to pass `native: true` or
`find_program()` would just ignore the
entire list and use whatever is in the config.

Please allow to write a search list to the
toolchain config.
Please don't drop the entire search list
from `find_program()` if 1 of the components
is `strip`. Maybe `find_program()` should
be diverted only if it has just 1 component,
and when is has the list then `native: true`
should be in effect?

**Comment by bonzini** [comment]
Related to your other issue with native: true/native: false, this shows how you really really want multiple platforms.

Specifying multiple binaries makes no sense since Meson wouldn't know which one to use.

**Comment by stsp** [comment]
> Specifying multiple binaries makes no sense since Meson wouldn't know which one to use.

The one that actually exists?
This has really little relevance to
my other problems: this ticket
was about the search lists. I.e.
on different distros the names
may differ. Like this:
i686-linux-gnu-strip, i386-elf-strip,
x86_64-linux-gnu-strip.
We need to pick up the one that
is actually present.

**Comment by bonzini** [comment]
You are confusing the purpose of a machine file. You do not do any detection with a machine file, as it's defining variables for things that could otherwise be detected (esp. via `find_program`).

I think what you want , at least as of now, is something like QEMU, which does all the detection and coordination on builds for multiple targets in a shell script, with meson/ninja invoked by configure/make. I can't exclude that meson will support this scenario in a standalone manner one day, but for now it is out of the scope of what meson is trying to be, let alone what meson is trying to do.

**Comment by stsp** [comment]
> You do not do any detection with a machine file, as it's defining variables for things that could otherwise
> be detected (esp. via find_program).

So could you please explain how
exactly is to "define" a cross-strip
in a machine file, if the exact name
may vary?

> but for now it is out of the scope of what meson is trying to be, let alone what meson is trying to do.

OMG, is it written anywhere, what
meson is trying to be/do? Maybe if
I had such reading beforehand, I
won't even waste my time on it at all.
But for now I just think it doesn't
support cross-compilation, while
it should.

**Comment by bonzini** [comment]
> So could you please explain how exactly is to "define" a cross-strip
in a machine file, if the exact name may vary?

Machine files are written for a specific setup. So the name won't vary for the one setup that the machine file is written for.

> But for now I just think it doesn't support cross-compilation, while it should.

It definitely doesn't support multi target compilation, and it doesn't support detecting cross compilers. But it does support cross compilation with handwritten (or generated by another program) cross files.

**Comment by eli-schwartz** [comment]
> Good, so I can write multiple machine files to cover various setups, is this right?
> But if so - how to select the right one to use? Or is this exactly what you propose to do with the configure script?

How would a configure script know what to use? Let's say you do indeed, as you say, have `i386-elf`, `i686-linux-gnu`, and `x86_64-linux-gnu` versions of your toolchain installed on the same machine.

And you run `./configure` to configure your build. You want an x86_64 build, but `./configure` notices that you have i386-elf installed so it tries to use that instead. Is that really a good idea? Is that even what a configure script does?

I don't think so! Configure scripts will tend to enforce a particular `--host` and `--build` to perform all detection.

**Comment by bonzini** [comment]
When I wrote multiple targets I meant the traditional GNU triple of build/host/target, where both host and target components are installed. Meson lacks support for that.

I am pretty familiar with your needs, since we both work on emulators. :)

**Comment by stsp** [comment]
That's exactly why I was
pondering meson actually.
I thought if its good enough
for qemu, I should certainly
try it too. :)
Currently I am a bit confused,
because it seems like it is not
fully suitable for qemu. In
which case why you choose it?

**Comment by bonzini** [comment]
Meson does not do all that QEMU needs, but it does the most complex part (configuring and building the emulator binaries) really well: more easily, with fewer lines of code and with fewer bugs. Basically QEMU's configure/make puts together all the different parts of the build, and Meson (thousands of lines of it, a lot more than the remaining configure/make code) is used for the biggest part which is why you'll see things like "QEMU moved to Meson".

This setup leaves pretty much these tasks to configure(*):
* It parses options on the command line (mostly using meson introspection data, so this code is rarely touched)
* It finds all the cross compilers (it has some fancy stuff for native compilation too but it isn't strictly necessary)
* It puts the results in a bunch of makefile includes as well as into meson native/cross files
* It invokes meson 

The split worked well for QEMU because the steps above are well isolated and rarely modified, so the result is overall easier to understand and modify than what we had before. And since QEMU already had a configure/make setup, we just had to decide when it made sense to stop moving stuff out of the configure script; there wasn't a lot of new shell and make code to be written.

(*) There's also some Python setup that can't be made in Meson because it uses Python, but it's irrelevant to you and a small thing overall.

**Comment by eli-schwartz** [comment]
Since as previously pointed out the usage model here should really be about producing the correct machine file, not checking for multiple possible providers of a machine file, as noted here:

> Its a bit of misunderstanding. My example lists x86_64-linux-gnu-xxx only because these tools accept `--32` flag, which makes them behave like i686-linux-gnu-xxx ones.
> 
> So varied is not only the name, but ALSO the needed flags. And the combination of particular name AND flags makes them entirely equal and interchangible.

I think it would make more sense to frame this as, how can one detect the correct machine file configuration to use for a given build.

Meson does have `meson env2mfile` which detects the current settings from PATH lookups and environment variables etc. It would not be unreasonable, I think, if `meson env2mfile --host=i386` knew how to detect an x86_64 prefixed toolchain and probe to see if --32 (-m32?) was available and viable.

**Comment by stsp** [comment]
For me the things are very
different. Its quite difficult
with configure+make (unless
you use automake) to get
the out-of-tree builds and
parallel builds rightly. And
if you moved to meson partially,
you still have to implement
both things in a makefile...
Plus I have lots of small
sub-projects, all of them
having cross/native parts,
so I'd need this configure/meson
split in all sub-projects. Also
most of these sub-projects
use custom or patched
compilers... I knew meson
can build the needed deps
with a wrap, and nope, not
when its a compiler. :(

So seems like you got quite
lucky with qemu to not fall
into these problems.

**Comment by stsp** [comment]
> It would not be unreasonable, I think, if meson env2mfile --host=i386 knew how to detect an
> x86_64 prefixed toolchain and probe to see if --32 (-m32?) was available and viable.

Hmm, sounds like a plan!
Is there a better documentation
for env2mfile than this one:
https://mesonbuild.com/Commands.html
I'd really like to familiarize myself
with that feature before commenting
on that topic.

As for --32 vs -m32 - seems -m32
is only viable for the compiles, while
--32 is for binutils.
Also if we take clang, it has a
unique set of options, like --target etc.
This all must be known to meson
of course.

**Comment by eli-schwartz** [comment]
> Hmm, sounds like a plan!  Is there a better documentation for env2mfile than this one: https://mesonbuild.com/Commands.html
>
> I'd really like to familiarize myself with that feature before commenting on that topic.

The sources are here: https://github.com/mesonbuild/meson/blob/master/mesonbuild/scripts/env2mfile.py

It is primarily / started off as, an integration of Debian's internal `debcrossgen` tool which they used in dpkg as part of the build rules for meson-based packages. There's a lot of room for people to help guide what it *should* be. :)

**Comment by stsp** [comment]
So from what I can see from
a quick look, is that it mainly
just translate the deb-specific
variables into a meson representation,
w/o allowing for any variability.
Maybe if you start with adding
llvm support to that command,
it will be more clear how to add
the variadic stuff? Because the
llvm tools are very much
incompatible with binutils
wrt cross-compilation. So once
you add that, the whole thing
would be already much feature-rich
than now. :)

Or should this be actually a
different tool, so to not being
bound to any deb-specific vars?
Maybe current tool does its work
very well, and needs no variadic
things for debian, so why not
to add a new one. :) With searching
capabilities, with clang support
and all that.

**Comment by eli-schwartz** [comment]
Not exactly. The debian handling is that env2mfile --debarch will ask debian's `dpkg-architecture` which ${triple} prefixes to scan $PATH for, but we should allow that via --host or --triple as well.

In general there's no need to add a new tool, if the existing tool could take an *option* instead.

**Comment by stsp** [comment]
Ok I thought `DEB_HOST_ARCH_OS`
and alike are the env vars, but it
appears the dict-alike output of
dpkg-architecture.
So yes, seems like you only need
to add --host and make it look for
a compiler into CC variable and
for linker type (bfd/lld) in some
other var?

In this case, with --host, CC and
LD_TYPE you can basically set up
any desired combo, and it will
automatically figure out what is
installed and what flags is to use?

**Comment by eli-schwartz** [comment]
That sounds about right, I think.

**Comment by stsp** [comment]
Would be cool, although I wonder
how difficult is to implement that.
Seems it would duplicate a lot of
what meson itself does, like
find_program(). So whether its
going to be a separate tool, or
some extension to platform files
syntax (which the currently existing
env2mfile could also use when
needed), is an interesting question.
Syntax extensions is what I actually
proposed in the beginning of the
thread. But of course extending
the tool is another option.
